lib/evmone/instructions_storage.cpp

```diff
@@ -0,0 +1,135 @@
+// evmone: Fast Ethereum Virtual Machine implementation
+// Copyright 2019 The evmone Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "instructions.hpp"
+
+namespace evmone::instr::core
+{
+namespace
+{
+struct StorageCostSpec
+{
+    bool net_cost;        ///< Is this net gas cost metering schedule?
+    int16_t warm_access;  ///< Storage warm access cost, YP: G_{warmaccess}
+    int16_t set;          ///< Storage addition cost, YP: G_{sset}
+    int16_t reset;        ///< Storage modification cost, YP: G_{sreset}
+    int16_t clear;        ///< Storage deletion refund, YP: R_{sclear}
+};
+
+constexpr auto storage_cost_spec = []() noexcept {
+    std::array<StorageCostSpec, EVMC_MAX_REVISION + 1> tbl{};
+
+    // Legacy cost schedule.
+    for (auto rev : {EVMC_FRONTIER, EVMC_HOMESTEAD, EVMC_TANGERINE_WHISTLE, EVMC_SPURIOUS_DRAGON,
+             EVMC_BYZANTIUM, EVMC_PETERSBURG})
+        tbl[rev] = {false, 200, 20000, 5000, 15000};
+
+    // Net cost schedule.
+    tbl[EVMC_CONSTANTINOPLE] = {true, 200, 20000, 5000, 15000};
+    tbl[EVMC_ISTANBUL] = {true, 800, 20000, 5000, 15000};
+    tbl[EVMC_BERLIN] = {
+        true, instr::warm_storage_read_cost, 20000, 5000 - instr::cold_sload_cost, 15000};
+    tbl[EVMC_LONDON] = {
+        true, instr::warm_storage_read_cost, 20000, 5000 - instr::cold_sload_cost, 4800};
+    tbl[EVMC_PARIS] = tbl[EVMC_LONDON];
+    tbl[EVMC_SHANGHAI] = tbl[EVMC_LONDON];
+    tbl[EVMC_CANCUN] = tbl[EVMC_LONDON];
+    return tbl;
+}();
+
+
+struct StorageCost
+{
+    int16_t gas_cost;
+    int16_t gas_refund;
+};
+
+// Build constexpr cost lookup table for SSTORE.
+constexpr auto sstore_costs = []() noexcept {
+    std::array<std::array<StorageCost, EVMC_STORAGE_MODIFIED_RESTORED + 1>, EVMC_MAX_REVISION + 1>
+        tbl{};
+
+    for (size_t rev = EVMC_FRONTIER; rev <= EVMC_MAX_REVISION; ++rev)
+    {
+        auto& e = tbl[rev];
+        if (const auto c = storage_cost_spec[rev]; !c.net_cost)  // legacy
+        {
+            e[EVMC_STORAGE_ASSIGNED] = {c.reset, 0};
+            e[EVMC_STORAGE_ADDED] = {c.set, 0};
+            e[EVMC_STORAGE_DELETED] = {c.reset, c.clear};
+            e[EVMC_STORAGE_MODIFIED] = {c.reset, 0};
```

Comment:
This could also be
```suggestion
            e[EVMC_STORAGE_MODIFIED] = e[EVMC_STORAGE_ASSIGNED];
```
because they are same cases for legacy.

Reply:
Yes, maybe even the other way around.

Reply:
I made it `e[EVMC_STORAGE_ASSIGNED] = e[EVMC_STORAGE_MODIFIED]`.